_posts/2021-03-14-Julia-1.6-what-has-changed-since-1.0.md

```diff
@@ -0,0 +1,737 @@
+[StagedFunctions.jl](https://github.com/NHDaly/StagedFunctions.jl) relaxes the restrictions on normal `@generated` functions so that they are also safe from [#265](https://github.com/JuliaLang/julia/issues/265)-like problems.
+[Tricks.jl](https://github.com/oxinabox/Tricks.jl/) uses this feature to make `hasmethod`, etc. resolve at compile-time, and then get updated if and when new methods are defined.
+This can allow for defining traits like _"anything that defines a `iterate` method"_.
+
+
+## Base and Standard Libraries
+
+### Threading
+Julia has full support for threading now.
+Not just the limited `@threads for` loops, but full Go-style threads.
+They are tightly integrated with the existing Async/Task/Coroutine system.
+In effect, threading works by unsetting the sticky flag on a `Task`, so that it is allowed to run on any thread.
+This is normally done via the `Threads.@spawn` macro, which replaces the `@async` macro.
+
+Interestingly, the `@threads for` macro still remains, and doesn't actually use much of the new machinery. It still uses the old way which is a bit tighter if the loop durations are almost identical.
+But the new threading stuff is fast, on the order of microseconds to send work off to another thread.
+Even for uses of `@threads for`, we get some wins from the improvements.
+`IO` is now thread-safe; `ReentrantLock` was added and is the kind of standard lock that you expect to exist.
+It has notifications on waiting work, etc.; and a big one: `@threads for` can now be nested without things silently being wrong.
+
+A lot of this actually landed in Julia 1.2, but Julia 1.3 was the release we think of as being for threading, as it gave us `Threads.@spawn`.
+
+Also, in Julia 1.6, we now have `julia -t auto` to start Julia with 1 thread per (logical) core.
+No more having to remember to set the `JULIA_NUM_THREADS` environment variable before starting it.
+
+### 5-arg `mul!`: in-place generalized multiplication and addition
+The 5 arg `mul!(C, A, B, α, β)` performs the operation equivalent to: `C .= A*B*α + C*β`, where `α`, `β` are scalars and `A`, `B` and `C` are compatible mixes of scalars, matrices and vectors.
+I am still of the opinion that it should have been called `muladd!`.
+This is the human-friendly version of `BLAS.gemm!` (i.e. GEneralized Matrix Multiplication) and it's ilk, thus the name `mul!`.
+It promises to alway compute the in-place mul-add in the most efficient, correct way for any `AbstractArray` subtype.
+In contrast, `BLAS.gemm!` computes the same thing, but with a bunch of conditions.
+It must be a strided array containing only BLAS scalars, and if one of the inputs is conjugated/transposed you need to input it in non-conjugated/transposed form, and then tell `BLAS.gemm!` via passing in `'C'` or `T` rather than `N`.
```

Comment:
```suggestion
It must be a strided array containing only BLAS scalars, and if one of the inputs is conjugated/transposed you need to input it in non-conjugated/transposed form, and then tell `BLAS.gemm!` via passing in `C` or `T` rather than `N`.
```